kpatch-build/create-diff-object.c

```diff
@@ -205,9 +205,20 @@ static void kpatch_bundle_symbols(struct kpatch_elf *kelf)
 	}
 }
 
+static void kpatch_add_child_symbol(struct symbol *parent, struct symbol *child)
+{
+	parent->nb_children++;
+	parent->children = realloc(parent->children,
+				   sizeof(*parent->children) * parent->nb_children);
+	child->parent = parent;
+	parent->children[parent->nb_children - 1] = child;
+}
+
 /*
  * During optimization gcc may move unlikely execution branches into *.cold
- * subfunctions. kpatch_detect_child_functions detects such subfunctions and
+ * subfunctions. Some functions can also be split into multiple *.part
+ * functions.
+ * kpatch_detect_child_functions detects such subfunctions and
```

Comment:
So the initial motivation for this change was a failure on the test gcc-static-local-var-5. The failure was due to a static local variable "lock" in function audit_log_end() being moved in the patched object into audit_log_end.part.17 .

One particular thing in the patched object is that both the audit_log_end and audit_log_end.part.* symbols exists, while the original only has audit_log_end. So the original audit_log_end is twined with the patched audit_log_end and static local references are only looked in the patched section related to audit_log_end and kpatch misses the fact that a reference was moved to the .part. section.

So, maybe the "child" concideration is not correct. But there needs to be a link between .part. functions and their original symbol (when it is also present on the patched object, as most of the time I've noticed it isn't) and the static variable correlation needs to be able to follow that link when looking for relocations.

Should I create a separate list of "parts" of a symbol? Other suggestions?

Reply:
Hm.  Does the patched audit_log_end() also reference the "lock" static local variable?  IIRC, "part" functions are clones of the originals.  So it seems like audit_log_end() should also reference the static local, unless I'm missing something.

Reply:
No, it doesn't have a reference to the lock. audit_log_end() almost only has a jump to audit_log_end.part in this case.

Reply:
Ah... I think I had a fundamental misunderstanding of this optimization.  Looking at gcc/ipa-split.c in the GCC source confirms that.  I had been thinking this was more like the constprop and isra optimizations.

So I take back my original comment!  It's very similar to the .cold optimization and I think the parent/child relationship makes sense after all.

Reply:
Thanks for having a look!

So how do you suggest I proceed with this PR? Split it in two, one with the test addition and patch modifications and the other with the create-diff-object changes?

Any suggestion for which one should be base on the other?

Reply:
Since we don't have a CI running these yet, I think I'd suggest doing the integration tests first, in one PR.  That way the other fixes can refer to them.

Reply:
Makes sense, I'll try sending that later today after rebasing on master and checking there aren't new issues!